src/store/transactional.ts

```diff
@@ -0,0 +1,126 @@
+import { fromJS, isCollection, isMap, List, Map } from "immutable";
+import { Ok, Result } from "ts-results";
+
+type Primitive = boolean | number | bigint | string | null | undefined | symbol;
+
+type Prefix<P, T extends any[]> = [P, ...T];
+type First<T extends any[]> = T extends Prefix<infer F, any[]> ? F : never;
+type Shift<T extends any[]> = T extends Prefix<any, infer R> ? R : [];
+
+// god type
+type Lens<T, P extends PropertyKey[]> =
+  P extends Prefix<any, any[]>
```

Comment:
😨 Typescript black magic

Reply:
are we going to be using any optics libraries?